enhancements/workload-partitioning/wide-availability-workload-partitioning.md

```diff
@@ -0,0 +1,731 @@
+metadata:
+  name: openshift-node-workload-partitioning-custom
+spec:
+  cpu:
+    isolated: 2-3
+    reserved: 0,1
+    # New addition
+    workloads:
+      - Infrastructure
+```
+
+To support upgrades and maintain better signaling for the cluster, the
+Performance Profile Controller will also inspect the Nodes to update a global
+identifier at start up. We will only update the identifier to `AllNodes` if we
+are running in Single Node and our Node has the capacity resource
+(`management.workload.openshift.io/cores`) for our 4.11 -> 4.12 upgrades. This
+should not be needed after 4.12 for all clusters. This should have no baring on
+4.11 HA/3NC clusters as this feature will not be back ported.
+
+### Workflow Description
+
+The end user will be expected to provide the default machine configs to turn on
+the feature for the whole cluster. As well as provide a `PerformanceProfile`
+manifest that describes their desired `isolated` and `reserved` CPUSet and the
+`Infrastructure` enum provided to the list in the `workloads` enum list.
+
+**High level sequence diagram:**
+
+Install Time Sequence
+
+```mermaid
+sequenceDiagram
+Alice->>Installer: Generate Manifests Install Config With CPU Partitioning
+Installer->>Alice: Create Dir with Cluster Manifests
+loop Apply
+    Alice->>Alice: Add PerformanceProfile Manifest
+end
+Alice->>Installer: Install Config With CPU Partitioning
+loop Generate
+    Installer->>Installer: Set Infrastructure Status
+end
+Installer-->>APIServer: Create Cluster
+APIServer-->>MCO: Get Infrastructure Status
+loop Generate
+    MCO->>MCO: Generate MC Configs
+end
+MCO-->>APIServer: Apply New MC
+loop Generate
+    MCO->>MCO: Generate Rendered MC
+end
+MCO-->>Node: Configure and Restart Node
+loop Apply
+    Node->>Node: Set Kubelet Config
+    Node->>Node: Set CRI- Config
+    Node->>Node: Kubelet Advertises Cores
+end
+Node-->>MCO: Finished Restart
+loop Generate
+    NTO->>NTO: Machine Config from PerformanceProfile
+end
+NTO-->>MCO: Apply Machine Config
+MCO-->>Node: Configure and Restart Node
+loop Apply
+    Node->>Node: Set Kubelet Config
+    Node->>Node: Set CRI- Config
+end
+Node-->>MCO: Finished Restart
+```
+
+- **Alice** is a human user who creates an Openshift cluster.
+- **Installer** is assisted installer that applies the user manifest.
+- **MCO** is the machine config operator.
+- **Node** is the kubernetes node.
+
+1. Alice sits down and uses the installer to generate the manifests with the
+   InstallConfig specifying `cpuPartitioningMode: AllNodes`
+   - `openshift-install create manifests`
+2. The installer generates the manifests to create the cluster, with the new
+   `Infrastructure.Status.CPUPartitioning: AllNodes`
+3. Alice adds the desired PerformanceProfile manifest for workload partitioning
+   to the `openshift` folder that was generated by the installer.
+4. Alice then creates the cluster via the installer.
+5. The installer will apply the manifests and during the bootstrapping process
```

Comment:
We do not have bootstrap procedure for NTO itself. That means the PerfProfile is not expanded to the necessary MachineConfigs at this time.
There was an attempt to do this in the past, but it was never finished.

On SNO this might not matter as the APi server and other critical pods will restart together with the node reboot. On 3NC this might not be as easy as the update goes node by node and so you will have nodes with and without partitioning in the cluster at the same time. Your proposed node admission hook might complain about that for example.

You also must expect failures (say rpm-ostree fails to deploy the kernel-rt or a customer MC conflicts) and recovery of a degraded intermediate state must be possible for all possible scenarios.

So I am actually missing the risk, failure and recovery description here. Normally it would not be such an issue, but workload partitioning is irreversible.

Reply:
Updated with more info on risks and mitigations.